### PR TITLE
Fix ui.update to correctly honor exclude_filetypes and include_buftypes

### PR DESCRIPTION
**Pull request by braxtons12**
This fixes the implementation of ui.update to correctly honor exclude_filetypes and include_buftypes so that other winbars aren't affected by barbecue.
The original implementation here could trample on other winbars in windows/buffers that shouldn't have been affected by barbecue, causing them to either be hidden, have weird behavior, or otherwise show incorrectly, because exclude_filetypes and include_buftypes were being checked in the update call, not the scheduled function to actually update the winbar.

One example of the previous conflict is with NeoTree's source_selector, which even when "neo-tree" was added to exclude_filetypes, would be broken by barbecue.nvim (it would only appear when a window other than the tree was focused, and only if the focus had changed at least once since the tree was opened).

### CODE REVIEW

**Comment by utilyre** [comment]
This patch makes [octo.nvim](https://github.com/pwntester/octo.nvim)'s buffer have a winbar (it shouldn't since it has a buftype of `acwrite`).

Could you give me a minimal config in which barbecue conflicts with neo-tree? I might be able to fix this without any regressions.

**Comment by braxtons12** [comment]
Here's the minimal config.
https://pastebin.com/SG6SWut7

This could be easily reworked to do the check before the winbar is built, which would solve the performance problem. I honestly just didn't think of that.

**Comment by braxtons12** [comment]
Moved the checks buftype, filetype, and visibility checks so they occur before the winbar is built. Solves the performance regression.

**Comment by utilyre** [comment]
> Here's the minimal config. https://pastebin.com/SG6SWut7

Can't reach the website. Could you post it here or open an issue since there's already a minimal config template there.

File: lua/barbecue/ui.lua [hunk 1]
```diff
@@ -147,6 +131,19 @@ function M.update(winnr)
     then
       return
     end
+      
+    if
+      not vim.tbl_contains(config.user.include_buftypes, vim.bo[bufnr].buftype)
+      or vim.tbl_contains(config.user.exclude_filetypes, vim.bo[bufnr].filetype)
+      or vim.api.nvim_win_get_config(winnr).relative ~= ""
+    then
+      return
```
Comment (utilyre) [inline lua/barbecue/ui.lua:140]: Is there a particular reason that you've removed these lines?

```lua
vim.wo[winnr].winbar = state.get_last_winbar(winnr)
state.clear(winnr)
```

**Comment by braxtons12** [comment]
> > Here's the minimal config. https://pastebin.com/SG6SWut7
> 
> Can't reach the website. Could you post it here or open an issue since there's already a minimal config template there.

Done.

> Is there a particular reason that you've removed these lines?
> 
> ```lua
> vim.wo[winnr].winbar = state.get_last_winbar(winnr)
> state.clear(winnr)
> ```

These lines specifically are what breaks the winbar from Neotree. If you shouldn't be changing the winbar to begin with, there's no reason to try to restore it back to some previous state (which is a race condition): just leave it alone. A consequence of that is that buftype/filetype check then needs to move into the scheduled function (as I've done in this PR) to make sure we don't enable the barbecue winbar for the wrong buftypes/filetypes

**Comment by braxtons12** [comment]
Fixed with [cbc4481](https://github.com/utilyre/barbecue.nvim/commit/cbc448149f9b3760a73fe6f50db7d224b40cdd95), so this is no longer needed

**Pull request closed without merging**